chain/chain/src/state_transition_data.rs

```diff
@@ -0,0 +1,265 @@
+use std::collections::hash_map::Entry;
+
+use std::collections::HashMap;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_chain_primitives::error::Error;
+use near_primitives::block::Block;
+use near_primitives::checked_feature;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{BlockHeight, ShardId};
+use near_primitives::utils::{get_block_shard_id, get_block_shard_id_rev};
+use near_store::db::STATE_TRANSITIONS_METADATA_KEY;
+use near_store::{DBCol, StorageError};
+use once_cell::unsync::OnceCell;
+
+use crate::{Chain, ChainStore, ChainStoreAccess};
+
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Default, Debug)]
+struct StateTransitionsMetadata {
+    /// Represents max not-yet-GCed height for each shard
+    start_heights: HashMap<ShardId, BlockHeight>,
+}
+
+impl Chain {
+    pub(crate) fn garbage_collect_state_transition_data(&self, block: &Block) -> Result<(), Error> {
+        let protocol_version =
+            self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+        if cfg!(feature = "shadow_chunk_validation")
+            || checked_feature!("stable", StatelessValidationV0, protocol_version)
+        {
+            let chain_store = self.chain_store();
+            let final_block_hash = *block.header().last_final_block();
+            if final_block_hash == CryptoHash::default() {
+                return Ok(());
+            }
+            let final_block = chain_store.get_block(&final_block_hash)?;
+            let final_block_height = final_block.header().height();
+            let final_block_present_chunks = final_block
```

Comment:
quick question, why do we need to filter on is_new_chunk? Why can't we just use the height create of the chunk in the function?

Reply:
The loop below would then look something like this:
```
for shard_id in all_shards:
    start_height = .....
    for height in start_height..final_block_height:
        delete DBCol::StateTransitionData
```

Reply:
yeah, good idea, it seems like using `height_created` makes it a bit simpler
updated in d0b14fc1d310ac02343bdc54fa07e4cf97e3a26e